relentless/potential/potential.py

```diff
@@ -385,20 +410,65 @@ def derivative(self, pair, param, r):
         ------
         ValueError
             If any value in `r` is negative.
+        TypeError
+            If the parameter with respect to which to take the derivative
+            is not a :py:class:`Variable`.
 
         """
         params = self.coeff.evaluate(pair)
         r,deriv,scalar_r = self._zeros(r)
         if any(r < 0):
             raise ValueError('r cannot be negative')
+        if not isinstance(var, Variable):
+            raise TypeError('Parameter with respect to which to take the derivative must be a Variable.')
 
-        # only evaluate at points inside [rmin,rmax], if specified
         flags = np.ones(r.shape[0], dtype=bool)
-        if params['rmin'] is not False:
-            flags[r < params['rmin']] = False
-        if params['rmax'] is not False:
-            flags[r > params['rmax']] = False
-        deriv[flags] = self._derivative(param, r[flags], **params)
+
+        for p in self.coeff.params:
+            # skip shift parameter
+            if p == 'shift':
+                continue
+
+            # try to take chain rule w.r.t. variable first
+            p_obj = self.coeff[pair][p]
+            if isinstance(p_obj, DependentVariable):
+                dp_dvar = p_obj.derivative(var)
+            elif isinstance(p_obj, IndependentVariable) and var is p_obj:
+                dp_dvar = 1.0
+            else:
+                dp_dvar = 0.0
+
+            # skip when dp_dvar is exactly zero, since this does not contribute
+            if dp_dvar == 0.0:
+                continue
+
+            # now take the parameter derivative
+            if p=='rmin':
+                # rmin deriv
+                flags = r < params['rmin']
+                deriv[flags] = -self._force(params['rmin'], **params)*dp_dvar
+            if p=='rmax':
+                # rmax deriv
+                if params['shift']:
+                    flags = r <= params['rmax']
+                    deriv[flags] += self._force(params['rmax'], **params)*dp_dvar
+                else:
+                    flags = r > params['rmax']
+                    deriv[flags] += -self._force(params['rmax'], **params)*dp_dvar
+            else:
+                # regular parameter derivative
+                below = np.zeros(r.shape[0], dtype=bool)
+                if params['rmin'] is not False:
+                    below = r < params['rmin']
+                    deriv[below] += -self._force(params['rmin'], **params)
+                above = np.zeros(r.shape[0], dtype=bool)
+                if params['rmax'] is not False:
+                    above = r > params['rmax']
+                    deriv[above] += -self._force(params['rmax'], **params)
+                flags = np.logical_and(~below, ~above)
+                deriv[flags] += self._derivative(p, r[flags], **params)*dp_dvar
+                if params['shift']:
+                    deriv[flags] -= -self._force(params['rmax'], **params)
```

Comment:
Same thing here. Also, make sure to add an error check (as in the `energy()` method) that it is an error to have `shift` be True and `rmax` not set. It will be a similar if-elif.